atomix/cluster/src/test/java/io/atomix/raft/protocol/ControllableRaftServerProtocol.java

```diff
@@ -99,6 +108,25 @@ private void send(
       final CompletableFuture responseFuture) {
     final var message = new Tuple<>(requestHandler, responseFuture);
     messageQueue.computeIfAbsent(memberId, m -> new LinkedList<>()).add(message);
+    addTimeOut(responseFuture);
+  }
+
+  private void addTimeOut(final CompletableFuture responseFuture) {
+    if (responseFuture != null) {
+      timeoutQueue.put(responseFuture, currentTime + requestTimeoutMillis);
+    }
+  }
+
+  public void tick(final long timeoutMillis) {
+    currentTime += timeoutMillis;
+    timeoutQueue.forEach(
+        (future, deadline) -> {
+          if (currentTime >= deadline) {
+            if (!future.isDone()) {
+              future.completeExceptionally(new TimeoutException());
+            }
```

Comment:
:wrench: I assume we also want to remove here the future from the timeout map?

Reply:
It's not strictly necessary because they are removed on receive but as there is no way to directly use/inspect the timeout map I agree that we can safely remove them here.

Reply:
But what if no receive is called due to timeout?

Reply:
:+1: